overlord/snapstate/storehelpers.go

```diff
@@ -664,17 +674,40 @@ func storeUpdatePlanCore(
 			return updatePlan{}, fmt.Errorf("internal error: snap %q not found", sar.InstanceName())
 		}
 
-		// TODO:COMPS: handle components here
+		currentComps, err := snapst.CurrentComponentInfos()
+		if err != nil {
+			return updatePlan{}, err
+		}
+
+		compNames := make([]string, 0, len(currentComps))
+		for _, comp := range currentComps {
+			compNames = append(compNames, comp.Component.ComponentName)
+		}
+
+		// TODO:COMPS: handle components losing a resource that is currently
+		// installed
+		compTargets, err := componentTargetsFromActionResult(sar, compNames)
+		if err != nil {
+			return updatePlan{}, fmt.Errorf("cannot extract components from snap resources: %w", err)
+		}
+
+		// if we still have no channel here, this means that we refreshed
+		// by-revision without specifying a channel. make sure we continue to
+		// track the channel that the snap is currently on
+		channel := up.RevOpts.Channel
+		if channel == "" {
+			channel = snapst.TrackingChannel
+		}
```

Comment:
should this code be a mini helper method on RevOpts ? do we do it only here and below?